Show notice in Course Editor to focus on Course Outline block or Insert one if block does not exist for first time users

Imran92 opened the pull request:
Resolves https://github.com/Automattic/sensei/issues/7331

## Proposed Changes
New users often don't know where they can add lessons, or sometimes, they don't know about the need of Course Outline block. So here, we've added a notice in the Course Editor that is shown to only those users who have not created any course yet.

Clicking on the link on that notice will focus (and scroll to, if necessary) on the Course Outline block if there is already one added in the editor.

If the Course Outline block is not there in the editor, it'll create one for the user. If at least one of the lessons in the Course Outline block is published, this notice will disappear.

To not bother the existing users, we've taken some measures. If the user has authored any course, this notice isn't shown to them. Once the user dismisses the notice, it's not shown to them anymore.

One thing to be noted is, that the dismissed flag is stored in localStorage. So the user may see it in a different browser or device. But it may not be very likely because once the user is in the Course editor, and playing around with it, there's a high probability that they'll have authored at least a course. So I didn't want to add another user meta for it as it may turn out to be not very useful.

I'll add some more tests for testing all the edge cases either in this PR, or a new one. I'm temporarily moving on to another issue.

## Testing Instructions
<!--
Add as many details as possible to help others reproduce the issue and test the changes.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

1. Create a new User who can create course
2. Create a new Course
3. Check that you see the desired notice on top
4. Click on the offline outline (Start with blank) or generate lessons with AI (Generate with AI)
5. Click on the `add some lessons` link on the notice
6. Make sure it takes the focus to the Course Outline block
7. Now delete all content from the editor
8. Now insert a pattern from Block Inserter (+ button on Top Toolbar of GB) -> Patterns -> Sensei LMS -> (Select any long pattern)
9. Now click on the 'add some lessons' link in the notice again
10. Make sure it scrolls you to the Course Outline block
11. Now remove everything again, or at least the Course Outline block
12. Click on the aforementioned link
13. Make sure it adds a Course Outline block in the editor
14. Now click on a lesson, from its toolbar, click 'Edit Lesson', and in the lesson editor, publish the lesson. Remember you are publishing only the lesson, not the course
15. Go back to the Course editor
16. Make sure the notice isn't there anymore
17. Create another similar user
18. Create a course. This time in the editor, dismiss the notice
19. Refresh the window, make sure the notice isn't appearing anymore
20. Now log in with an user who has authored at least one course
21. Make sure you don't see the notice for the user

https://github.com/Automattic/sensei/assets/6820724/cf3c8b30-0306-4229-930f-dfb3ba8eed56

## Pre-Merge Checklist
<!-- Complete applicable items on this checklist **before** merging. Items that are not applicable can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed. -->
- [x] PR title and description contain sufficient detail and accurately describe the changes
- [x] Acceptance criteria is met
- [x] Decisions are publicly documented
- [x] Adheres to coding standards ([PHP](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/), [JavaScript](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/), [CSS](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/css/), [HTML](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/html/))
- [x] All strings are translatable (without concatenation, handles plurals)
- [x] Follows our naming conventions (P6rkRX-4oA-p2)
- [ ] Hooks (p6rkRX-1uS-p2) and functions are documented
- [x] New UIs are responsive and use a [mobile-first approach](https://zellwk.com/blog/how-to-write-mobile-first-css/)
- [x] New UIs match the designs
- [x] Different user privileges (admin, teacher, subscriber) are tested as appropriate
- [x] Code is tested on the minimum supported PHP and WordPress versions
- [x] User interface changes have been tested on the latest versions of Chrome, Firefox and Safari
- [ ] "Needs Documentation" label is added if this change requires updates to documentation
- [ ] Known issues are created as new GitHub issues

Imran92 committed:
Show notice to create outline, or focus outline first time

Imran92 committed:
Add import comments

Imran92 committed:
Add notice file in webpack

Imran92 committed:
Enqueue first course notice only when first course

Imran92 committed:
Test published lesson checking function

Imran92 committed:
Add tests for conditionally creating or selecting outline block

Imran92 committed:
Update logic to show notice

Imran92 committed:
Add changelog

Imran92 committed:
Update the comment

donnapep commented:
Hey @Imran92! Did a super quick test.

1. It looks like the notice is displayed as soon as I go into the editor. It should only be displayed as per this criteria:
> Notice is triggered when a user creates their **first** course and **after** selecting a course layout.

2. Also, I'm reconsidering this criteria:

> Notice is no longer displayed as soon as the course contains at least one published lesson

I'm thinking perhaps we should change it to:

> Notice is no longer displayed as soon as the course contains at least one lesson, regardless of status

If I saw that notice that asked me to add some lessons, and I did that, I'd be confused as to why the notice was still there. Then, I think we could catch the issue with draft lessons by tweaking this frontend notice and adding a CTA (though as part of a separate issue and PR):

![Screenshot 2023-12-07 at 4 08 11 PM](https://github.com/Automattic/sensei/assets/1190420/7844819e-595e-4fc7-a6b0-cd7e650f3168)

3. I think this criteria hasn't been met, but perhaps it would also benefit from some tweaking:

> Notice is not displayed for subsequent courses the user may create.

I created 2 courses and didn't add lessons to either of them, but I saw the notice in both. After I published a lesson in the first course, I still saw the notice in the second course. I'm happy to go with whichever of these is easiest to implement:

- Implement the criteria as described; that is, tie the notice to the first course the user creates and _only_ to the first course the user creates.
- Leave the notice on both, but don't show it on _any_ course after lessons have been added to either of them.

Please let me know your thoughts on all of this. 🙇🏻‍♀️

Imran92 committed:
Only show notice after sensei pattern inserted

Imran92 committed:
Show notice only when no course of any status exists

Imran92 committed:
Merge branch 'trunk' into add/notice-to-create-course-outline-block-first-time

Imran92 committed:
Merge branches 'add/notice-to-create-course-outline-block-first-time' and 'add/notice-to-create-course-outline-block-first-time' of https://github.com/Automattic/sensei into add/notice-to-create-course-outline-block-first-time

Imran92 committed:
Move pattern variable check inside function

Imran92 committed:
Show notice when there is no published lesson

Imran92 commented:
Thanks for checking it @donnapep ! 

> 1. It looks like the notice is displayed as soon as I go into the editor. It should only be displayed as per this criteria:

I've updated it here https://github.com/Automattic/sensei/pull/7364/commits/ddb29ea77be3018b029a1a019a6fc91d29705dce, now the notice will only be shown after a pattern is selected

> If I saw that notice that asked me to add some lessons, and I did that, I'd be confused as to why the notice was still there

Agreed, I'm updating it here https://github.com/Automattic/sensei/pull/7364/commits/4bc987c2729b41b0e308c4e98730e829e4ea5d68 to show the notice only when there are no lessons in it.

> Notice is not displayed for subsequent courses the user may create.

Oh, it was only considering if you've published a course. I've updated the logic here https://github.com/Automattic/sensei/pull/7364/commits/f4246699d13712ae0f0613b8f7226a1bfa26eb60 to consider any Course of any status. Showing it the first time sounds better as it helps with having less notice cluttering on the top.

>  I think we could catch the issue with draft lessons by tweaking this frontend notice and adding a CTA (though as part of a separate issue and PR):

Absolutely agree with this one. After seeing the deep dive, I think any improvement here that helps with publishing the course and lessons properly is worth doing.

Imran92 committed:
Fix test

donnapep committed:
Merge branch 'trunk' into add/notice-to-create-course-outline-block-first-time

donnapep commented:
@Imran92 It's better, but still isn't quite working as expected for me. The notice is now showing after a course layout is selected, but if I save the course without adding any lessons, and then come back in later, the notice isn't there anymore.

Also, looks like some tests are broken.

Imran92 committed:
Merge branch 'trunk' into add/notice-to-create-course-outline-block-first-time

Imran92 committed:
Fix failing tests

Imran92 committed:
Show notice only when no lessons are created

Imran92 committed:
Wait for pattern on new course, else show right away

Imran92 commented:
Thanks for taking a look @donnapep !

I've updated the logic here https://github.com/Automattic/sensei/pull/7364/commits/e84c62afe6a4a705631ad5370ee49045d1d02cde to load the notice scripts only when the user has at least a lesson created under a course.

Here, I've gone with the following approach, as it required less checking with db query

> Leave the notice on both, but don't show it on any course after lessons have been added to either of them.

I've left some edge case behaviors, but wanted to surface them here to get your opinion on them -

- **User has created lessons without associating with any courses**: If the user has created any lesson that are not associated with any course, then we still show the notice on Courses
- **User has deleted the course after creating**: If the user deletes only the Course after creating a Course with lessons, we're not doing any special handling for that.
- **User has deleted the lessons after creating**: If the user creates a Course with lessons, but then deletes the lessons, we don't keep a track of that
- **User saves Course without adding any lesson, but then closes and reopens**: As per our need, when the user creates a new Course, we show the notice only after the user has selected a pattern. But if the user decides to save the Course without any lesson then closes the window, and opens the same course again to edit it, we show the notice right away. Because we don't show the pattern picker modal again. And if the user has already selected a pattern the last time, it's not straightforward to detect if a pattern was used (we can keep a meta, or flag in storage, but I wasn't sure the extra work to cover this edge case was necessary). So as there no modal to block the viewport, and if the user is eligible of seeing the notice, we show the user the notice right away when they open a saved course for editing later.

Thanks!

donnapep commented:
@Imran92 Thanks! The logic looks good. I had a bit of confusion when testing. At one point, the notice stopped showing for new users. :scream: After stepping through code, I realized it was because I had previously dismissed the notice, which I see is tracked in local storage. Is this how WordPress typically handles client-side notices? I know for server-side it's usually an option in the `wp_options` table. But if so, I think it would be helpful to mention it in the PR description as I think it could fool others too. 🙂

An FYI that I haven't done a code review; I've only tested the behavior & logic. Given our discussion about not really needing this for when a user selects a pattern from the block inserter, if there is any feedback from a code review that requires reworking the patterns code, please just remove that logic. It's not part of the flow that new users are most likely going to move through, and so is not the flow we're most interested in capturing. No sense in spending any more time there. 🙂 I'm also hoping this notice will be temporary until we're able to secure a design resource to help us with this. Thx!

renatho commented on assets/js/admin/first-course-creation-notice.js:
```diff
@@ -0,0 +1,181 @@
+ */
+import { getFirstBlockByName } from '../../blocks/course-outline/data';
+
+export const hasOutlineBlock = () =>
```
Would `getOutlineBlock` be more clear here?

Imran92 replied:
Yap, name update here https://github.com/Automattic/sensei/pull/7364/commits/3d82cd56848d37ac0f63e2876b96b6062e00915d. Thanks!

renatho commented on assets/js/admin/first-course-creation-notice.js:
```diff
@@ -0,0 +1,181 @@
+		const { insertBlock } = dispatch( 'core/block-editor' );
+
+		courseOutlineBlock = createBlock( 'sensei-lms/course-outline' );
+		await insertBlock( courseOutlineBlock );
```
I didn't test it, but I think the selection happens by default given the [documentation](https://developer.wordpress.org/block-editor/reference-guides/data/data-core-block-editor/#insertblock) (`updateSelection` defaults to `true`).

If confirmed, I think it doesn't need to be an `async` function, and we can add the following selection in an `else`. WDYT?

Imran92 replied:
Makes sense, I've updated the code here https://github.com/Automattic/sensei/pull/7364/commits/a633f08b144fc19edff010461f62dc208903bd7c

renatho replied:
Thank you @Imran92!
With this, I think the function doesn't need to be `async` anymore, right?

Imran92 replied:
Right! Removed here https://github.com/Automattic/sensei/pull/7364/commits/3d47a014d0417186f835b323403c52cf04104a77, thanks @renatho

renatho commented on assets/js/admin/first-course-creation-notice.js:
```diff
@@ -0,0 +1,181 @@
+	let noticeRemoved = false;
+
+	const notice = __(
+		'Nice! Now you can <a href="javascript:;" onclick="window?.handleCourseOutlineBlockIncomplete();">add some lessons</a> to your course.',
```
My suggestion wouldn't match the design, but it would match the default behavior of the Gutenberg notices.

Should we tweak it to use a separate button, being the `actions` from the Gutenberg notice instead? cc @donnapep

So we wouldn't need to put the function in the `window`, we wouldn't need to add the `onclick` and `href="javascript:;"` as HTML attributes, it would get more organized in terms of code, and the users would the same behavior of other notices.

If we decide to not change it, we need to extract the HTML and logic from the translation string.

Imran92 replied:
If we're good with the design change, we can go for it 👍 Our current design of showing it as a URL looks better to me. But as showing it as button is simpler, as that's the default behavior. I think the main benefit of this will be not having to use the `__unstableHTML` property.

donnapep replied:
I'm fine with changing it to use a button if it helps simplify things. In that case, we could change the message to:

> Nice! Now you can add some content to your course. Add Lessons

where _Add Lessons_ is a button that jumps to / adds the Course Outline block.

Let's see what that looks like. 🙂

donnapep replied:
Updated comment [here](https://github.com/Automattic/sensei/pull/7364#issuecomment-1881521558) 🙂

renatho reviewed (commented):
In general, it looks good.

But I'd like to propose something. IMO, this is a very small helper which doesn't worth such complexity to maintain. My suggestion would be:

- Don't worry about persisting the dismiss, since it would simplify the code, and it's a helper to "jump" to the Course Outline for new courses.
- Since we won't display the notice when having draft lessons, create the notice as soon as the wizard is closed (skipping earlier or choosing a pattern), and remove the notice when:
  - Clicking on "add some lessons".
  - When clicking on the "X".
  - When clicking on the course outline actions "Start with blank" or "Generate with AI".

With that, we don't need to:
- Monitor the dispatch.
- Subscribe the store.
- Have any backend logic.
- Have new JS assets (we can have a file imported in the `course-edit.js`).
- Worry about persistency.
- Worry if the post is new or not.

WDYT? cc @donnapep

Imran92 committed:
change name of hasOutlineBlock function

Imran92 committed:
Add meta for new post for course

Imran92 committed:
Remove inserter pattern selection detection code

Imran92 committed:
Manually select outline block only when available

Imran92 committed:
Remove unused import

Imran92 committed:
Merge branch 'trunk' into add/notice-to-create-course-outline-block-first-time

Imran92 commented:
Thanks for checking @donnapep!

> Is this how WordPress typically handles client-side notices? I know for server-side it's usually an option in the wp_options table. But if so, I think it would be helpful to mention it in the PR description as I think it could fool others too. 🙂

Yap you are right, if it's a global notice to be shown once, it's saved in `wp_options`, and for per-user notices, the flag is saved in `wp_usermeta`. I didn't find a convention for handing this in the frontend. I've shared my thoughts in the PR description for it- 
`One thing to be noted is, that the dismissed flag is stored in localStorage. So the user may see it in a different browser or device. But it may not be very likely because once the user is in the Course editor, and playing around with it, there's a high probability that they'll have authored at least a course. So I didn't want to add another user meta for it as it may turn out to be not very useful.`
But if we think it's necessary to save it as meta and it can have other usage too or can be important data, we can add it.

> It's not part of the flow that new users are most likely going to move through, and so is not the flow we're most interested in capturing

Thanks! I've removed it here https://github.com/Automattic/sensei/pull/7364/commits/ee75e87c1776006b50ab176c0d4573275743c8e2 .

Imran92 commented:
Thanks @renatho for your nice suggestion and the new approach you suggested, I'm good with having a version with a bit of a smaller set of requirements for it. My only concern is I'm not sure users who are familiar with the Sensei environment (from the perspective of our current set of requirements, they are users who have created any lesson that belongs to any course) would like to see a notice on top. And also, even though we have an extra asset, it's loaded only for the specific case of the first Course. A little benefit that I like about this is we load less code if all subsequent usages. But I'm okay to load it as part of `course-edit.js` too, no problem!

Thanks again! <3

Imran92 committed:
Remove unnecessary async

Imran92 committed:
Set default value to null

Imran92 committed:
Set default to false

Otherwise php above 8 throws an error in testing. So we manually set the _new_post to true for new post. If we do it in backed, it causes the new course modal to appear for existing courses

donnapep commented:
> Thanks @renatho for your nice suggestion and the new approach you suggested, I'm good with having a version with a bit of a smaller set of requirements for it. My only concern is I'm not sure users who are familiar with the Sensei environment (from the perspective of our current set of requirements, they are users who have created any lesson that belongs to any course) would like to see a notice on top.

After thinking about this for a bit, if Renatho's suggestions https://github.com/Automattic/sensei/pull/7364#pullrequestreview-1794598738 significantly simplify the logic, I'm open to displaying this notice for existing users too. I would want to tweak the messaging though:

> There are no lessons in your course. Add some now"

where "Add some now" is a button that jumps to / adds the Course Outline block.

donnapep commented:
We're pausing this work until we get through some of the other onboarding improvements.

Imran92 commented:
Maybe we can close this as we have the Course tour now?
